File: scripts/kobuki_simulator.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from geometry_msgs.msg import Twist, Pose, Quaternion, Vector3, Point, PoseWithCovarianceStamped, TransformStamped
from nav_msgs.msg import Odometry
import tf2_ros
from tf_transformations import euler_from_quaternion, quaternion_from_euler
import threading
import numpy as np

class KobukiSimulator( Node ):

  def __init__( self, initial_x = 1.0, initial_y = 1.0, initial_yaw = 0.0 ):
    super().__init__( 'kobuki_simulator' )
    self.get_logger().info( 'Initializing Kobuki Simulator' )
    self.declare_parameter( 'initial_x', rclpy.Parameter.Type.DOUBLE )
    self.declare_parameter( 'initial_y', rclpy.Parameter.Type.DOUBLE )
    self.declare_parameter( 'initial_yaw', rclpy.Parameter.Type.DOUBLE )

    self.initial_x = initial_x
    try:
      self.initial_x = self.get_parameter( 'initial_x' ).get_parameter_value().double_value
    except rclpy.exceptions.ParameterUninitializedException:
      pass

    self.initial_y = initial_y
    try:
      self.initial_y = self.get_parameter( 'initial_y' ).get_parameter_value().double_value
    except rclpy.exceptions.ParameterUninitializedException:
      pass

    self.initial_yaw = initial_yaw
    try:
      self.initial_yaw = self.get_parameter( 'initial_yaw' ).get_parameter_value().double_value
    except rclpy.exceptions.ParameterUninitializedException:
      pass

    self.real_pose_publish_rate = 5.0 # [Hz]
    self.simulate_ground_friction = True
    self.reset = False

    # Odometry pose covariace
    self.cov_x = 1e-05
    self.cov_y = 1e-05
    self.cov_z = 1e-03

    pose_position = Point( x = self.initial_x, y = self.initial_y, z = 0.0 )
    q = quaternion_from_euler( 0.0, 0.0, self.initial_yaw )
    pose_quaternion = Quaternion( x = q[0], y = q[1], z = q[2], w = q[3] )
    self.current_pose = Pose( position = pose_position, orientation = pose_quaternion )

    self.current_speed = Twist( linear = Vector3( x = 0.0, y = 0.0, z = 0.0 ),
                                angular = Vector3( x = 0.0, y = 0.0, z = 0.0 ) )

    self.odom_broadcaster = tf2_ros.TransformBroadcaster( self )
    self.create_subscription( Twist, '/cmd_vel', self.move, 1 )
    self.create_subscription( String, '/active', self.velocity_state, 1 )
    self.create_subscription( Pose, '/initial_pose', self.set_initial_pose, 1 )
    self.create_subscription( PoseWithCovarianceStamped, '/initialpose', self.set_initial_pose_with_cov, 1 )
    self.pub_odom = self.create_publisher( Odometry, '/odom', 1 )
    self.pub_real_pose = self.create_publisher( Pose, '/real_pose', 1 )

  def set_initial_pose( self, initial_pose ):
    if initial_pose.position.x == float('inf') and \
       initial_pose.position.y == float('inf'):
      q = quaternion_from_euler( 0.0, 0.0, self.initial_yaw )
      initial_pose = Pose( position = Point( x = self.initial_x, y = self.initial_y, z = 0.0 ),
                           orientation = Quaternion( x = q[0], y = q[1], z = q[2], w = q[3] ) )
      self.reset = True
    self.current_pose = initial_pose
    self.pub_real_pose.publish( self.current_pose )

  def set_initial_pose_with_cov( self, initialpose ):
    self.current_pose = initialpose.pose.pose
    self.pub_real_pose.publish( self.current_pose )

  def update_real_pose( self, vx, vy, vyaw, dt ):
    roll, pitch, yaw = euler_from_quaternion( ( self.current_pose.orientation.x,
                                                self.current_pose.orientation.y,
                                                self.current_pose.orientation.z,
                                                self.current_pose.orientation.w ) )

    delta_x = (vx * np.cos( yaw ) - vy * np.sin( yaw )) * dt
    delta_y = (vx * np.sin( yaw ) + vy * np.cos( yaw )) * dt
    if self.simulate_ground_friction:
      delta_yaw = 0.9 * vyaw * dt
    else:
      delta_yaw = vyaw * dt

    x = self.current_pose.position.x + delta_x
    y = self.current_pose.position.y + delta_y
    yaw = yaw + delta_yaw

    q = quaternion_from_euler( roll, pitch, yaw )
    self.current_pose = Pose( position = Point( x = x, y = y, z = 0.0 ),
                              orientation = Quaternion( x = q[0], y = q[1], z = q[2], w = q[3] ) )
    self.pub_real_pose.publish( self.current_pose )

  def publish_odom( self, x, y, yaw, vx, vy, vyaw, current_time ):
    t = TransformStamped()
    t.header.stamp = current_time.to_msg()
    t.header.frame_id = 'odom'
    t.child_frame_id = 'base_link'

    t.transform.translation.x = x
    t.transform.translation.y = y
    t.transform.translation.z = 0.0

    q = quaternion_from_euler( 0.0, 0.0, yaw )
    t.transform.rotation.x = q[0]
    t.transform.rotation.y = q[1]
    t.transform.rotation.z = q[2]
    t.transform.rotation.w = q[3]

    self.odom_broadcaster.sendTransform( t )

    # next, we'll publish the odometry message over ROS
    odom = Odometry()
    odom.header.stamp = current_time.to_msg()
    odom.header.frame_id = 'odom'
    odom.child_frame_id = 'base_link'

    # set the position
    odom.pose.pose = Pose( position = Point( x = x, y = y, z = 0.0 ),
                           orientation = Quaternion( x = q[0], y = q[1], z = q[2], w = q[3] ) )

    odom.pose.covariance = [ self.cov_x, 0.0, 0.0, 0.0, 0.0, 0.0,
                             0.0, self.cov_y, 0.0, 0.0, 0.0, 0.0,
                             0.0, 0.0, 10000000.0, 0.0, 0.0, 0.0,
                             0.0, 0.0, 0.0, 10000000.0, 0.0, 0.0,
                             0.0, 0.0, 0.0, 0.0, 10000000.0, 0.0,
                             0.0, 0.0, 0.0, 0.0, 0.0, self.cov_z ]

    # set the velocity
    odom.twist.twist = Twist( linear = Vector3( x = vx, y = vy, z = 0.0 ),
                              angular = Vector3( x = 0.0, y = 0.0, z = vyaw ) )

    odom.twist.covariance = [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                              0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                              0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                              0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                              0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                              0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ]

    self.pub_odom.publish( odom )

  def main_loop( self ):
    x = 0.0
    y = 0.0
    yaw = 0.0
    # Depends on incoming speed
    current_time = self.get_clock().now()
    last_time = self.get_clock().now()
    count = 1
    rate = self.create_rate( self.real_pose_publish_rate )
    try:
      while rclpy.ok():
        if self.reset:
          x, y, yaw = 0.0, 0.0, 0.0
          self.reset = False
        vx, vy, vyaw = self.get_current_speed()
        current_time = self.get_clock().now()
        dt = (current_time - last_time).nanoseconds * 1e-9

        self.update_real_pose( vx, vy, vyaw, dt )

        delta_x = (vx * np.cos( yaw ) - vy * np.sin( yaw )) * dt
        delta_y = (vx * np.sin( yaw ) + vy * np.cos( yaw )) * dt
        if self.simulate_ground_friction:
          delta_yaw = 0.9 * vyaw * dt
        else:
          delta_yaw = vyaw * dt

        x += delta_x
        y += delta_y
        yaw += delta_yaw

        # publish the message every 1 [s]
        #if count >= int( self.real_pose_publish_rate ):
        #  self.publish_odom( x, y, yaw, vx, vy, vyaw, current_time )
        #  count = 0
        #count += 1
        self.publish_odom( x, y, yaw, vx, vy, vyaw, current_time )

        last_time = current_time
        rate.sleep()
    except KeyboardInterrupt:
      self.get_logger().info( 'Shutting down kobuki_simulator' )
      pass

  def move( self, twist ):
    if np.isnan( twist.linear.x ) or np.isnan( twist.angular.z ):
      self.get_logger().warn( 'Invalid speed command received: lin.x: %s, ang.z: %s' % (str(twist.linear.x), str(twist.angular.z)) )
      return
    # movement is restricted to x and yaw
    twist.linear.y = 0.0
    twist.linear.z = 0.0
    twist.angular.x = 0.0
    twist.angular.y = 0.0
    self.current_speed = twist

  def velocity_state( self, state ):
    if state.data == 'idle':
      self.current_speed = Twist( linear = Vector3( x = 0.0, y = 0.0, z = 0.0 ),
                                  angular = Vector3( x = 0.0, y = 0.0, z = 0.0 ) )

  def get_current_speed( self ):
    return self.current_speed.linear.x, self.current_speed.linear.y, self.current_speed.angular.z

  def shutdown( self ):
    # stop turtlebot
    self.get_logger().info( 'Stopping Kobuki Simulator' )
    # a default Twist has linear.x of 0 and angular.z of 0.  So it'll stop TurtleBot
    self.current_speed = Twist( linear = Vector3( x = 0.0, y = 0.0, z = 0.0 ),
                                angular = Vector3( x = 0.0, y = 0.0, z = 0.0 ) )
    # sleep just makes sure TurtleBot receives the stop command prior to shutting down the script
    rospy.sleep( 1 )


if __name__ == '__main__':
  rclpy.init()
  kobuki_sim = KobukiSimulator()

  thread = threading.Thread( target = rclpy.spin, args = (kobuki_sim, ), daemon = True )
  thread.start()

  kobuki_sim.main_loop()

  rclpy.shutdown()
  thread.join()

File: src/kobuki_simulator.cpp
#include "kobuki_simulator.h"

#include <tf/transform_datatypes.h>
#include <tf/transform_broadcaster.h>
#include <geometry_msgs/TransformStamped.h>

using namespace std;


KobukiSimulator::KobukiSimulator( float initial_x,
                                  float initial_y,
                                  float initial_yaw )
: m_initial_x( initial_x ),
  m_initial_y( initial_y ),
  m_initial_yaw( initial_yaw ),
  m_simulate_ground_friction( true ),
  m_reset( true )
{
  ROS_INFO( "Initializing Kobuki Simulator" );

  m_node_handle.param<float>( "/kobuki_simulator/initial_x", m_initial_x, initial_x );
  m_node_handle.param<float>( "/kobuki_simulator/initial_y", m_initial_y, initial_y );
  m_node_handle.param<float>( "/kobuki_simulator/initial_yaw", m_initial_yaw, initial_yaw );

  m_current_pose.position.x = m_initial_x;
  m_current_pose.position.y = m_initial_y;
  m_current_pose.position.z = 0;
  geometry_msgs::Quaternion quat = tf::createQuaternionMsgFromYaw( m_initial_yaw );
  m_current_pose.orientation = quat;

  m_cmd_vel_sub = m_node_handle.subscribe( "cmd_vel", 1, &KobukiSimulator::move, this );
  m_active_sub = m_node_handle.subscribe( "cmd_vel_active", 1, &KobukiSimulator::velocity_state, this );
  //m_cmd_vel_sub = m_node_handle.subscribe( "yocs_cmd_vel_mux/output/cmd_vel", 1, &KobukiSimulator::move, this );
  //m_active_sub = m_node_handle.subscribe( "yocs_cmd_vel_mux/active", 1, &KobukiSimulator::velocity_state, this );
  m_initial_pose_sub = m_node_handle.subscribe( "initial_pose", 1, &KobukiSimulator::set_initial_pose, this );
  m_initial_pose_with_cov_sub = m_node_handle.subscribe( "initialpose", 1, &KobukiSimulator::set_initial_pose_with_cov, this );
  m_odom_pub = m_node_handle.advertise<nav_msgs::Odometry>( "odom", 10 );
  m_real_pose_pub = m_node_handle.advertise<geometry_msgs::Pose>( "real_pose", 1 );
}

void
KobukiSimulator::set_initial_pose( const geometry_msgs::Pose::ConstPtr& initial_pose )
{
  if( initial_pose->position.x == INFINITY || initial_pose->position.y == INFINITY )
  {
    m_current_pose.position.x = m_initial_x;
    m_current_pose.position.y = m_initial_y;
    m_current_pose.position.z = 0;
    geometry_msgs::Quaternion quat = tf::createQuaternionMsgFromYaw( m_initial_yaw );
    m_current_pose.orientation = quat;
    m_reset = true;
  }
  else
  {
    m_current_pose.position = initial_pose->position;
    m_current_pose.orientation = initial_pose->orientation;
  }
  ROS_INFO( "Initial pose received: (%.2f, %.2f)",
            m_current_pose.position.x,
            m_current_pose.position.y );
  m_real_pose_pub.publish( m_current_pose );
}

void
KobukiSimulator::set_initial_pose_with_cov( const geometry_msgs::PoseWithCovarianceStamped::ConstPtr& initialpose )
{
  m_current_pose = initialpose->pose.pose;
  ROS_INFO( "Initial pose received: (%.2f, %.2f)",
            m_current_pose.position.x,
            m_current_pose.position.y );
  m_real_pose_pub.publish( m_current_pose );
}

void
KobukiSimulator::update_real_pose( float vx, float vy, float vyaw, float dt )
{
  tf::Quaternion q( m_current_pose.orientation.x,
                    m_current_pose.orientation.y,
                    m_current_pose.orientation.z,
                    m_current_pose.orientation.w );
  tf::Matrix3x3 m( q );
  double roll, pitch, yaw;
  m.getRPY( roll, pitch, yaw );

  float delta_x = (vx * cos( yaw + m_initial_yaw ) - vy * sin( yaw + m_initial_yaw )) * dt;
  float delta_y = (vx * sin( yaw + m_initial_yaw ) + vy * cos( yaw + m_initial_yaw )) * dt;
  float delta_yaw = 0;
  if( m_simulate_ground_friction )
  {
    delta_yaw = 0.9 * vyaw * dt;
  }
  else
  {
    delta_yaw = vyaw * dt;
  }

  float x = m_current_pose.position.x + delta_x;
  float y = m_current_pose.position.y + delta_y;
  yaw = yaw + delta_yaw;

  m_current_pose.position.x = x;
  m_current_pose.position.y = y;
  m_current_pose.position.z = 0;
  geometry_msgs::Quaternion quat = tf::createQuaternionMsgFromYaw( yaw );
  m_current_pose.orientation = quat;

  m_real_pose_pub.publish( m_current_pose );
}

void
KobukiSimulator::publish_odom( float x, float y, float yaw, float vx, float vy, float vyaw, const ros::Time& current_time )
{
  geometry_msgs::TransformStamped odom_trans;
  odom_trans.header.stamp = current_time;
  odom_trans.header.frame_id = "odom";
  odom_trans.child_frame_id = "base_link";
  odom_trans.transform.translation.x = x;
  odom_trans.transform.translation.y = y;
  odom_trans.transform.translation.z = 0;
  geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw( yaw );
  odom_trans.transform.rotation = odom_quat;
  m_odom_broadcaster.sendTransform( odom_trans );

  nav_msgs::Odometry odom;
  odom.header.stamp = current_time;
  odom.header.frame_id = "odom";

  //set the position
  odom.pose.pose.position.x = x;
  odom.pose.pose.position.y = y;
  odom.pose.pose.position.z = 0.0;
  odom.pose.pose.orientation = odom_quat;

  //set the velocity
  odom.child_frame_id = "base_link";
  odom.twist.twist.linear.x = vx;
  odom.twist.twist.linear.y = vy;
  odom.twist.twist.angular.z = vyaw;

  //publish the message
  m_odom_pub.publish( odom );
}

void
KobukiSimulator::main_loop()
{
  float x = 0;
  float y = 0;
  float yaw = 0;
  float vx = 0;
  float vy = 0;
  float vyaw = 0;
  float delta_x = 0;
  float delta_y = 0;
  float delta_yaw = 0;
  float dt = 0;
  // Depends on incoming speed
  ros::Time current_time = ros::Time::now();
  ros::Time last_time = ros::Time::now();
  int count = 1;

  ros::Rate rate( kRealPosePublishRate );
  while( ros::ok() )
  {
    ros::spinOnce();
    if( m_reset )
    {
      x, y, yaw = 0, 0, 0;
      m_reset = false;
    }

    get_current_speed( vx, vy, vyaw );
    current_time = ros::Time::now();
    dt = (current_time - last_time).toSec();

    update_real_pose( vx, vy, vyaw, dt );

    delta_x = (vx * cos( yaw + m_initial_yaw ) - vy * sin( yaw + m_initial_yaw )) * dt;
    delta_y = (vx * sin( yaw + m_initial_yaw ) + vy * cos( yaw + m_initial_yaw )) * dt;
    if( m_simulate_ground_friction )
    {
      delta_yaw = 0.9 * vyaw * dt;
    }
    else
    {
      delta_yaw = vyaw * dt;
    }

    x += delta_x;
    y += delta_y;
    yaw += delta_yaw;

    /*
    // publish the message every 1 [s]
    if( count >= kRealPosePublishRate )
    {
      publish_odom( x, y, yaw, vx, vy, vyaw, current_time );
      count = 0;
    }
    count += 1;
    */
    publish_odom( x, y, yaw, vx, vy, vyaw, current_time );

    last_time = current_time;
    rate.sleep();
  }
}

void
KobukiSimulator::move( const geometry_msgs::Twist& twist )
{
  if( twist.linear.x == NAN || twist.angular.z == NAN )
  {
    ROS_WARN( "Invalid speed command received: lin.x: %f, ang.z: %f", twist.linear.x, twist.angular.z );
    return;
  }
  // movement is restricted to x and yaw
  m_current_speed.linear.x = twist.linear.x;
  m_current_speed.angular.z = twist.angular.z;
}

void
KobukiSimulator::velocity_state( const std_msgs::String& state )
{
  ROS_INFO( "Current subscriptor: %s", state.data.c_str() );
  if( "idle" == state.data )
  {
    m_current_speed.linear.x = 0;
    m_current_speed.angular.z = 0;
  }
}

void
KobukiSimulator::get_current_speed( float& vx, float& vy, float& vyaw )
{
  vx = m_current_speed.linear.x;
  vy = m_current_speed.linear.y;
  vyaw = m_current_speed.angular.z;
}


